Archive/Hairer.lean

```diff
@@ -0,0 +1,242 @@
+    HasCompactSupport f :=
+  HasCompactSupport.of_support_subset_isCompact (isCompact_closedBall 0 1) (support_subset f)
+
+end SmoothSupportedOn
+
+end normed
+open SmoothSupportedOn
+
+section missing_polynomial
+open MvPolynomial Submodule
+
+variable {R σ : Type*} [CommSemiring R] (n : ℕ)
+
+lemma restrictTotalDegree_eq_span {n : ℕ} :
+    restrictTotalDegree σ R n =
+    span R ((fun c : σ →₀ ℕ ↦ monomial c (1 : R)) '' {s : σ →₀ ℕ | s.sum (fun _ e ↦ e) ≤ n}) := by
+  ext P; constructor <;> intro h
+  · rw [← P.support_sum_monomial_coeff]
+    refine sum_mem fun c hc ↦ ?_
+    rw [← mul_one (coeff c P), ← smul_eq_mul, ← smul_monomial]
+    rw [mem_restrictTotalDegree] at h
+    exact smul_mem _ _ (subset_span <| mem_image_of_mem _ <| (le_totalDegree hc).trans h)
+  · refine span_le.mpr ?_ h
+    rintro x ⟨c, hc, rfl⟩
+    rw [SetLike.mem_coe, mem_restrictTotalDegree]
+    cases subsingleton_or_nontrivial R
+    · rw [Subsingleton.elim ((fun c ↦ monomial c 1) c) 0, totalDegree_zero]; apply zero_le
+    · rw [totalDegree_monomial _ one_ne_zero]; exact hc
+
+/- Can be obtained by combining `LinearMap.proj` and `MvPolynomial.evalₗ`. -/
+/-- The evaluation of a multivariate polynomial at a point, as a linear map. -/
+def evalAtₗ {R σ : Type*} [CommSemiring R] (x : σ → R) : MvPolynomial σ R →ₗ[R] R where
+  toFun P := eval x P
+  map_add' := by simp
+  map_smul' := by simp
+
+/- what we actually want -/
+lemma analyticOn_eval {R σ : Type*} [Fintype σ] [IsROrC R] (P : MvPolynomial σ R) :
+    AnalyticOn R (fun x ↦ eval (EuclideanSpace.equiv σ R x) P) univ := fun x _ ↦ by
+  apply P.induction_on (fun r ↦ ?_) (fun p q hp hq ↦ ?_) fun p i hp ↦ ?_ -- refine doesn't work
+  · simp_rw [eval_C]; exact analyticAt_const
+  · simp_rw [eval_add]; exact hp.add hq
+  · simp_rw [eval_mul, eval_X]; exact hp.mul ((ContinuousLinearMap.proj (R := R) i).analyticAt x)
+
+lemma finite_stuff' [Finite σ] (N : ℕ) : {s : Multiset σ | Multiset.card s ≤ N}.Finite := by
+  classical
+  have := Fintype.ofFinite σ
+  let S := N • (Finset.univ.val : Multiset σ)
+  apply Finset.finite_toSet (Multiset.toFinset (Multiset.powerset S)) |>.subset
+  intro s hs
+  rw [Set.mem_setOf] at hs
+  rw [Finset.mem_coe, Multiset.mem_toFinset, Multiset.mem_powerset, Multiset.le_iff_count]
+  intro x
+  simp only [S, Multiset.count_nsmul, Multiset.count_univ, mul_one]
+  exact le_trans (s.count_le_card x) hs
+
+lemma finite_stuff [Finite σ] (N : ℕ) : {s : σ →₀ ℕ | s.sum (fun _ e ↦ e) ≤ N}.Finite := by
+  classical
+  change {s : σ →₀ ℕ | s.sum (fun _ => id) ≤ N}.Finite
+  simp only [← Finsupp.card_toMultiset]
+  refine Set.Finite.of_finite_image ?_ (Multiset.toFinsupp.symm.injective.injOn _)
+  convert finite_stuff' (σ := σ) N
+  ext x
+  rw [← AddEquiv.coe_toEquiv, Set.mem_image_equiv]
+  simp
+
+instance [Finite σ] : Module.Finite R (restrictTotalDegree σ R n) := by
+  rw [Module.finite_def, fg_top, restrictTotalDegree_eq_span]
+  exact Submodule.fg_span ((finite_stuff _).image _)
```

Comment:
Let's try to golf these a bit. The instance could go into Mathlib.RingTheory.MvPolynomial.Basic together with `restrictTotalDegree_eq_span`.

Reply:
These are now golfed but I can't move the instance to RingTheory.MvPolynomial.Basic because of the lack of imports. #find_home suggests it has to stay in this file.